### PR TITLE
generate named slots content for stubbed components via shallowMount

### LINKED ISSUE #1307: Named slots' content is not generated for stubbed components via shallowMount
### What problem does this feature solve?
Higher-order components make use of named slots. It's possible to test higher-order components via shallowMount because named slot content can be mocked, but it's not possible to test components which consume higher-order components because their generated named slot output is not available in the stubbed content.

https://github.com/vuejs/vue-test-utils/pull/782 added support for outputting default slot content, but not named slots.

To be clear, an example of a component which needs this support is:

```
<!-- myComponent.vue -->
<template>
  <HigherOrderComponent foo="bar">
    default slot content
    <template v-slot="header">header slot content</template>
    <template v-slot="footer">footer slot content<HigherOrderComponent foo="waldo"/></template>
  </HigherOrderComponent>
</template>
```

Currently, this is stubbed out to only:

`<higherordercomponent-stub foo="bar">default slot content</higherordercomponent-stub>`

The end user experience should be that components which generate named slot content "just work" when testing with shallowMount.

### What does the proposed API look like?
The API would not change.

### PR DESCRIPTION
**Pull request by chriswa**
resolves #1307

### CODE REVIEW

**Comment by chriswa** [comment]
Hi! This is my first ever OSS contribution. Please let me know any feedback you may have! :)

**Comment by chriswa** [comment]
I did create an Issue for this PR, but I guess I failed to link it:
https://github.com/vuejs/vue-test-utils/issues/1307

**Comment by ndchelsea** [comment]
This is a super helpful fix... What's the chances of this being merged?

**Comment by GiacomoVoss** [comment]
Any updates on this? It's really helpful and would make snapshot testing with Vuetify components a lot better.

**Comment by lmiller1990** [comment]
Hey @chriswa , awesome first OSS contribution. I will review this now and give you feedback so w can get this merged.

File: packages/create-instance/create-component-stubs.js [hunk 1]
```diff
@@ -86,31 +86,48 @@ export function createStubFromComponent(
...
+            )
+          }
+        : {
+            ...this.$props
+          }
+
+      const slots = context ? context.slots() : this._renderProxy.$slots
+
+      // ensure consistent ordering of slots (default first, then alphabetical)
+      const sortedSlotEntries = Object.entries(slots)
```
Comment (lmiller1990) [inline packages/create-instance/create-component-stubs.js:144]: Nice usage of `Object.entries`

File: packages/create-instance/create-component-stubs.js [hunk 2]
```diff
@@ -86,31 +86,48 @@ export function createStubFromComponent(
...
+      sortedSlotEntries.sort(([slotNameA], [slotNameB]) =>
+        slotNameA === 'default'
+          ? -1
+          : slotNameB === 'default'
+          ? 1
+          : slotNameA.localeCompare(slotNameB)
```
Comment (lmiller1990) [inline packages/create-instance/create-component-stubs.js:150]: I would have used a simple `>` and `<` for the string compare. Not sure it makes a big difference - it is common to use characters like `á` in code? (my language does not have those characters, so I have never seen this before).

If we are doing to support using `localeCompare`, maybe we should have a test containing a `á` character vs an `a` character. Although probably not strictly necessary.

File: test/specs/shallow-mount.spec.js [hunk 3]
```diff
@@ -90,8 +91,40 @@ describeRunIf(process.env.TEST_ENV !== 'node', 'shallowMount', () => {
...
+        '    <child-stub></child-stub>\n' +
+        '  </template-stub>\n' +
+        '  <template-stub slot="header">Hello</template-stub>\n' +
+        '</child-stub>'
+    )
+  })
+
+  it('renders named slots for functional components', () => {
+    const localVue = createLocalVue()
+    localVue.component('child', {
```
Comment (lmiller1990) [inline test/specs/shallow-mount.spec.js]: nice job! it's easy to forget the case of functional components.

File: packages/create-instance/create-component-stubs.js [hunk 4]
```diff
@@ -86,31 +86,48 @@ export function createStubFromComponent(
   // ignoreElements does not exist in Vue 2.0.x
   if (Vue.config.ignoredElements) {
     Vue.config.ignoredElements.push(tagName)
+    if (Vue.config.ignoredElements.indexOf('template-stub') === -1) {
```
Comment (lmiller1990) [inline packages/create-instance/create-component-stubs.js:130]: Why is this necessary?

**Review by lmiller1990 — COMMENTED** [review]
This is an awesome first OSS contribution. Left a few comments, only one really requires any action (first one about ignored elements).

Might also be good to rebase on the last develop. Great job! I can't wait to get this one merged in.

**Comment by lmiller1990** [comment]
@chriswa left some comments, let me know what you think. Excited to get this one in.

File: packages/create-instance/create-component-stubs.js (hunk 2 above)
Comment (dobromir-hristov) [reply packages/create-instance/create-component-stubs.js:150]: Just a thought, this kind of ternary is super confusing. I would deff advise for some sort of reordering or other.

File: packages/create-instance/create-component-stubs.js (hunk 4 above)
Comment (dobromir-hristov) [reply packages/create-instance/create-component-stubs.js:130]: So the `template-stub` stays as `<template-stub />` in the finished render, correct?

File: packages/create-instance/create-component-stubs.js [hunk 5]
```diff
@@ -86,31 +86,48 @@ export function createStubFromComponent(
...
+      const sortedSlotEntries = Object.entries(slots)
+      sortedSlotEntries.sort(([slotNameA], [slotNameB]) =>
+        slotNameA === 'default'
+          ? -1
+          : slotNameB === 'default'
+          ? 1
+          : slotNameA.localeCompare(slotNameB)
+      )
+
+      const children = sortedSlotEntries.map(([slotName, slotChildren]) =>
```
Comment (dobromir-hristov) [inline packages/create-instance/create-component-stubs.js]: So this should render the default slot, and all others as children of the `template-stub`? Wont this be a super breaking change for snapshots?

Comment (lmiller1990) [reply packages/create-instance/create-component-stubs.js]: True, this will break all snapshots using the original behavior. We probably should have some snapshot tests in VTU as well to capture these regressions. But then we will need to move to Jest, I think? Or we can do a manual comparison between strings.

Probably not the biggest deal, we just need to make sure that's clear in release notes. Something like:

NOTE: You will need to update all snapshots when updating.

File: packages/create-instance/create-component-stubs.js [hunk 6]
```diff
@@ -80,25 +80,42 @@ function resolveOptions(component, _Vue) {
...
+//     ctx.$options.parent._vnode.data &&
+//     ctx.$options.parent._vnode.data.scopedSlots
+//   ) {
+//     const slotKeys: Array<string> = ctx.$options.parent._vnode.data.scopedSlots
+//     return keys(slotKeys).filter(x => x !== '_normalized' && x !== '$stable')
+//   }
+//
+//   return []
+// }
+
```
Comment (JessicaSachs) [inline packages/create-instance/create-component-stubs.js:118]: So @lmiller1990 I'm mostly concerned about why no tests break when I comment this out

File: packages/create-instance/create-component-stubs.js [hunk 7]
```diff
@@ -5,8 +5,8 @@ import {
   throwError,
   camelize,
   capitalize,
-  hyphenate,
-  keys
+  hyphenate
+  // keys
```
Comment (JessicaSachs) [inline packages/create-instance/create-component-stubs.js:9]: Commented out for formatting until we can figure out how to merge everything in.

File: packages/create-instance/create-component-stubs.js [hunk 8]
```diff
@@ -110,36 +127,53 @@ export function createStubFromComponent(
...
+      // if (context) {
+      //   children = sortedSlots
+      // } else if (this.$options._renderChildren) {
+      //   children = this.$options._renderChildren
+      // } else {
+      //   children = getScopedSlotRenderFunctions(this).map(x =>
+      //     this.$options.parent._vnode.data.scopedSlots[x]()
+      //   )
+      // }
+
```
Comment (JessicaSachs) [inline packages/create-instance/create-component-stubs.js:169]: This was initially how I expanded out and merged dev and this PR, but was failing so I commented it out and reverted to mainly this PR's version of rendering slots

File: test/specs/shallow-mount.spec.js [hunk 9]
```diff
@@ -130,7 +136,11 @@ describeRunIf(process.env.TEST_ENV !== 'node', 'shallowMount', () => {
     expect(wrapper.find({ name: 'Foo' }).exists()).to.equal(true)
     expect(wrapper.find('.new-example').exists()).to.equal(true)
     expect(wrapper.html()).to.equal(
-      '<foo-stub>\n' + '  <p class="new-example">text</p>\n' + '</foo-stub>'
+      '<foo-stub>\n' +
+        '  <template-stub slot="newSyntax">\n' +
```
Comment (JessicaSachs) [inline test/specs/shallow-mount.spec.js:140]: The need to update all the assertions with `template-stub` worry me. This will be a breaking change :/

**Comment by afontcu** [comment]
Hi! As mentioned [here](https://github.com/vuejs/vue-test-utils/issues/246#issuecomment-608739968), we're not merging this before v1 is released.

Just in case someone stumbles upon this PR – there are 2 workarounds:

1. use mount
2. write your own stub that implements the named slots


thanks for the effort put on this one, though! 🤗

**Comment by lmiller1990** [comment]
I thought about this a bit more, I'm not convinced this behavior is actually correct. I think children of stubbed components, including slots, should not be rendered. I am in favor of closing this as a wontfix.

We need to figure out what shallowMount should/shouldn't do.

**Comment by lmiller1990** [comment]
This looks to be a significant a breaking change at this point. I also am not sure this actually makes sense - I think stubbed children should not render their slots.

I'm going to close this one. Thanks for the work - sorry the team was unresponsive and this didn't get the attention it should have when the PR was opened.

We are looking to a v1 release soon, and we will revisit shallowMount and what it should be rendering for the Vue 3 integration.

**Comment by chenxeed** [comment]
Hi @lmiller1990 , just sharing my opinion here regarding your thought:

> I think children of stubbed components, including slots, should not be rendered.

I'm afraid if there's multiple interpretation of defining this use-case.

I agree that generally, `shallowMount` shouldn't render the child component and simply stub them.

However, slot is different. Slot is a way for the parent to "inject" the content they expect in the child component. The child is responsible for the placement of the slot, but it's the parent responsibility to ensure the content injected are working as expected.

Thus, testing the slot content is the parent responsibility.

For example, when we have a component that needs to load a modal component like [b-modal](https://bootstrap-vue.org/docs/components/modal) with specific content for the named slot:

```vue
// parent component
<template>
  <b-modal v-model="visible">
    <template #modal-header>
      {{ title }}
      <button @click="onClickModalHeader">
    </template>
</b-modal>
```

On testing the parent component, I don't care with how the `<b-modal>` renders the slot since I believe `b-modal` will do its job properly. I just want to make sure the content of the header are rendering the title that I expect and the button inside are working.

_note: example with jest_

```js
const parentComponentWrapper = shallowMount(parentComponent);

parentComponent.setData({ title: 'The Parent Title' });
const modal = parentComponent.findComponent({ name: 'BModal' });
expect(modal.html()).toContain('The Parent Title'); // better if can assert which slot it's expecting
modal.find('button').trigger('click')
// assert the side effect from `onClickModalHeader`
```

The idea is, slots are used by parent for the parent's logic, and we need to test them too. Using `mount` to test them will defeat the purpose of `shallowMount` to only render components from the parent layer.

I beg you to reconsider :bow:  and let me know if my example above doesn't makes sense or not applicable. Thank you :bow:

**Comment by dobromir-hristov** [comment]
This is taken care of in Vue 2. Default slots are rendered behind a flag, named slots only via a special helper. :)

**Comment by chenxeed** [comment]
Thanks to respond @dobromir-hristov 

> named slots only via a special helper

Can you clarify what's the special helper you mentioned about?

Is it the manual stub that @danieltian suggested:
```ts
const wrapper = shallowMount(RootComponent, {
  stubs: { ChildComponent }
}

expect(wrapper.find(ChildComponent).text()).toContain('some text');
```

**Comment by dobromir-hristov** [comment]
Sort of. In shallowMount default slots can be auto rendered, but not named and scoped slots. We cant know that data, as it is deeply buried in the render function tree of the components you are stubbing. Does it make sense?

I tried to make this for a while, bit there is just no sensible way to do render all slots, while still stubbing.

**Comment by chenxeed** [comment]
@lmiller1990 @dobromir-hristov  I've seen the discussion has been going on heavily in #1307 which has better explanation than mine. My appreciation to you and the core team for the stable release, and taking in the concern of this issue :heart: 

> Does it make sense?

Yes, seems like the issue here is more towards the implementation, not the goal behind.

I agree with @lmiller1990 [suggestion](https://github.com/vuejs/vue-test-utils/issues/1307#issuecomment-617503092) that this behavior can be made as optional.
I'll followup the discussion from #1307 and look forward in the next vue-test-utils.

Again, thank you! :1st_place_medal:
